### PR TITLE
Support db_datareader/db_datawriter

### PR DESCRIPTION
**Pull request by shalinilohia50**
### Description
Implement database roles `db_datareader` and `db_datawriter` 

* Members of the db_datareader fixed database role can read all data from all user tables and views. User objects can exist in any schema except sys and INFORMATION_SCHEMA.
* Members of the db_datawriter fixed database role can add, delete, or change data in all user tables. In most use cases, this role is combined with db_datareader membership to allow reading the data that is to be modified.

Engine PR: https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/98

### Issues Resolved
Task: [BABEL-3883](https://jira.rds.a2z.com/browse/BABEL-3883)
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

### Test Scenarios Covered ###
Test Plan: https://quip-amazon.com/xZyQABuVCz9u/Test-Plan-for-database-roles 
* **Use case based -** Added

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 1]
```diff
@@ -282,6 +306,9 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 	appendStringInfo(&query, "DROP OWNED BY dummy, dummy, dummy CASCADE; ");
 
 	/* Then drop db_accessadmin, db_owner and dbo in that order */
+	appendStringInfo(&query, "DROP ROLE dummy; ");
+	appendStringInfo(&query, "DROP ROLE dummy; ");
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/dbcmds.c:310]: Should we execute DROP OWNED BY? Will it be able to help cleaning up default privileges on these roles?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 2]
```diff
@@ -4271,6 +4285,10 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							foreach(lc, grant->grantees)
 							{
 								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								/* Special database roles should throw an error. */
+								if (rol_spec->rolename != NULL && (strcmp(rol_spec->rolename, db_datareader) == 0 || strcmp(rol_spec->rolename, db_datawriter) == 0))
+									ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+										errmsg("Cannot grant, deny or revoke permissions to or from special roles.")));
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Can we refactor this? It is duplicated at 3 places. And does it make sense to add a check for all the fixed db roles?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 3]
```diff
@@ -1846,6 +1847,22 @@ check_alter_role_stmt(GrantRoleStmt *stmt)
...
+	
+	/* Get the database owner. */
+	db_owner_name = get_owner_of_db(get_cur_db_name());
+
+	/*
+	 * Disallow ALTER ROLE if current login is not a member of sysadmin and
+	 * not the database owner.
+	 */
+	if (!has_privs_of_role(GetSessionUserId(), get_role_oid("sysadmin", false)) &&
+		!has_privs_of_role(GetSessionUserId(),get_role_oid(db_owner_name, false)))
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Only db_owner membership check on current_user should be sufficient. I'm not sure in existing why are we relying on the session user for add/dropping the membership of database roles.

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 4]
```diff
@@ -1263,23 +1296,152 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
...
+		return;
+	}
+
+	if (OidIsValid(get_role_oid(db_datawriter, true)))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("role \"%s\" already exists. Please drop the role and restart upgrade.", db_datawriter)));
+		return;
+	}
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/dbcmds.c]: Can we refactor it to make a common function for database_roles to throw this error? Because it seems redundant.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 2 above)
Comment (anju15bharti) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: +1

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 5]
```diff
@@ -2491,3 +2492,99 @@ get_owner_of_schema(const char *schema)
...
+
+	stmts = parsetree_nth_stmt(stmt_list, i++);
+	update_AlterDefaultPrivilegesStmt(stmts, schema, schema_owner, dbo_role, db_datareader, NULL);
+
+	stmts = parsetree_nth_stmt(stmt_list, i++);
+	update_AlterDefaultPrivilegesStmt(stmts, schema, schema_owner, dbo_role, db_datawriter, NULL);
+
+	PG_TRY();
+	{
+		SetUserIdAndSecContext(get_bbf_role_admin_oid(), save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:2567]: here as well we are setting our internal role, shouldn't it also escalate the priv here? @HarshLunagariya

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 3 above)
Comment (anju15bharti) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: +1

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 6]
```diff
@@ -249,30 +267,36 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
...
+	const char *db_datareader;
+	const char *db_datawriter;
 
 	dbo = get_dbo_role_name(dbname);
 	db_owner = get_db_owner_name(dbname);
 	db_accessadmin = get_db_accessadmin_role_name(dbname);
 	schema = get_dbo_schema_name(dbname);
 	guest_schema = get_guest_schema_name(dbname);
+	db_datareader = get_db_datareader_name(dbname);
+	db_datawriter = get_db_datawriter_name(dbname);
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/dbcmds.c:286]: Why have we casted it to const char* in many other places as well?

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 7]
```diff
@@ -1263,23 +1296,152 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
...
+						NULL);
+		}
+		tuple = heap_getnext(tblscan, ForwardScanDirection);
+	}
+
+	/* Cleanup. */
+	table_endscan(tblscan);
+	table_close(namespace_rel, RowExclusiveLock);
+	pfree(dbname);
+	pfree(query.data);
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/dbcmds.c:1395]: Why are we only pfreeing query.data and not query because we will always initialize the query with memory allocation in `initStringInfo`?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 8]
```diff
@@ -4227,16 +4235,17 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 			}
 		case T_GrantStmt:
 			{
-				GrantStmt *grant = (GrantStmt *) parsetree;
-				char	   *dbname = get_cur_db_name();
-				const char *current_user = GetUserNameFromId(GetUserId(), false);
+				GrantStmt	*grant = (GrantStmt *) parsetree;
+				char		*dbname = get_cur_db_name();
+				const char	*db_datareader = get_db_datareader_name(dbname);
+				const char	*db_datawriter = get_db_datawriter_name(dbname);
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: +1 , shouldn't we define these with char instead of const char* anf pfree it as well since in `get_db_datareader_name` and `get_db_datawriter_name` we are allocating the memory?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 9]
```diff
@@ -2491,3 +2492,99 @@ get_owner_of_schema(const char *schema)
...
+	if (schema_owner && strcmp(schema_owner, dbo_role) != 0 && strcmp(schema_owner, db_owner) != 0)
+	{
+		// do nothing
+	}
+	else
+	{
+		schema_owner = GetUserNameFromId(get_owner_of_schema(schema), false);
+	}
+
+	initStringInfo(&query);
```
Comment (anju15bharti) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c:2547]: we should pfree `query` as well.

**Review by anju15bharti — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 1 above)
Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/src/dbcmds.c:310]: Do we really need this? 
Adding DROP OWNED BY doesn't change the behavior but affects the pg_dump tests and throwing error:
```
 DROP DATABASE IF EXISTS babel_2170_db1;
 GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: permission denied to drop objects)~~
```

**Review by rishabhtanwar29 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 1 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:310]: What is it trying to drop?

File: contrib/babelfishpg_tsql/src/pltsql_utils.c [hunk 10]
```diff
@@ -2504,3 +2505,113 @@ get_owner_of_schema(const char *schema)
...
+{
+	if (rolname != NULL &&
+		(strcmp(rolname, rol1) == 0 ||
+		 strcmp(rolname, rol2) == 0 ||
+		 strcmp(rolname, rol3) == 0))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+			errmsg("Cannot grant, deny or revoke permissions to or from special roles.")));
+	}
+}
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pltsql_utils.c]: Shouldn't this be applicable to all the fixed db roles? In that case we can directly use the existing MACRO. so that everytime we won't need to change this function.

File: test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out [hunk 11]
```diff
@@ -46,3 +46,17 @@ CREATE TABLE master_dbo.tbl_creation_should_succeed (
...
+BEGIN 
+    IF EXISTS (SELECT 1 FROM
+information_schema.tables WHERE
+table_name = 'tbl_creation_should_succeed' AND
+table_schema = 'master_dbo') THEN
+        EXECUTE 'GRANT ALL ON
+master_dbo.tbl_creation_should_succeed TO master_dbo';
+    END IF;
+END $$;
+GO
```
Comment (HarshLunagariya) [inline test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out:62]: Why this change?

File: test/JDBC/input/test_windows_login-vu-prepare.mix [hunk 12]
```diff
@@ -1,4 +1,4 @@
--- sla 55000
+-- sla 70000
```
Comment (HarshLunagariya) [inline test/JDBC/input/test_windows_login-vu-prepare.mix:1]: Was it consistently failing? We should try to investigate the perf. degrade reasons.

File: test/JDBC/input/BABEL-SP_COLUMN_PRIVILEGES.mix [hunk 13]
```diff
@@ -1,5 +1,5 @@
--- sla 20000
--- sla_for_parallel_query_enforced 20000
+-- sla 50000
+-- sla_for_parallel_query_enforced 50000
```
Comment (HarshLunagariya) [inline test/JDBC/input/BABEL-SP_COLUMN_PRIVILEGES.mix]: Was it consistently failing? We should try to investigate the perf. degrade reasons.

File: test/JDBC/expected/datareader_datawriter-vu-verify.out [hunk 14]
```diff
@@ -0,0 +1,639 @@
+-- tsql
+-- bbf dump does not dump password so reset the password
```
Comment (HarshLunagariya) [inline test/JDBC/expected/datareader_datawriter-vu-verify.out:2]: is_member() and is_rolemember() test case should be part for upgrade test file.

+ Normal testcases to validate whether any other fixed role is not member of these roles.

File: test/JDBC/expected/datareader_datawriter.out [hunk 15]
```diff
@@ -0,0 +1,1830 @@
...
+<NULL>
+~~END~~
+
+
+SELECT IS_ROLEMEMBER('db_datawriter', 'db_roles_r1')
+GO
+~~START~~
+int
+<NULL>
+~~END~~
```
Comment (HarshLunagariya) [inline test/JDBC/expected/datareader_datawriter.out]: Shouldn't it be 0?

File: test/JDBC/expected/datareader_datawriter.out [hunk 16]
```diff
@@ -0,0 +1,1830 @@
...
+
+
+SELECT IS_MEMBER('db_datawriter')
+GO
+~~START~~
+int
+0
+~~END~~
+
+
```
Comment (HarshLunagariya) [inline test/JDBC/expected/datareader_datawriter.out:1520]: Can we refactor the test - to test positive/negative cases only when current user is only member of either db_datareader and db_datawriter? Currently, in all cases it is either member of both or none.

File: test/JDBC/expected/datareader_datawriter.out [hunk 17]
```diff
@@ -0,0 +1,1830 @@
...
+use master;
+go
+
+-- tsql user=db_roles_l1 password=123
+-- Basic membership check
+use db_roles_db1
+go
+
+SELECT IS_MEMBER('db_datareader')
+GO
```
Comment (HarshLunagariya) [inline test/JDBC/expected/datareader_datawriter.out:1624]: Can we please add testcases for following system procs?
    * sp_helpdbfixedrole testcase -- I think this got covered in existing test cases
    * sp_helpuser
    * sp_helprole
    * sp_helprolemember

**Review by HarshLunagariya — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 1 above)
Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/src/dbcmds.c:310]: Here we are dropping the db_datareader and db_datawriter roles. 
If there had to be any dependent privilege, then DROP ROLE <> wouldn't have been successful. Do we need DROP OWNED BY in that case?

File: test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out (hunk 11 above)
Comment (shalinilohia50) [reply test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out:62]: We created this table from the PG endpoint before upgrade. We execute some GRANTs and ALTER DEFAULT PRIVILEGES statements and to have that privilege on this particular object, we have to grant permission to master_dbo else we get permission denied while upgrade.

Comment (HarshLunagariya) [reply test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out:62]: Does it fail upgrade?

Comment (shalinilohia50) [reply test/JDBC/expected/BABEL-CHECK-CONSTRAINT-vu-prepare.out:62]: Yes, because during upgrade we execute those GRANT statements and that give permission denied on this object since this table is created in BBF schema via PG endpoint.

**Review by HarshLunagariya — COMMENTED** [review]
Overall changes looks good to me!, Just one testcase addition for is_rolemember() is required as discussed offline.

**Comment by shalinilohia50** [comment]
> Overall changes looks good to me!, Just one testcase addition for is_rolemember() is required as discussed offline.

Missed mentioning it earlier, that test case is already added in the `datareader_datawriter-vu-verify.mix` test file.

**Review by HarshLunagariya — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 1 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/dbcmds.c:310]: Why do we not use `DROP ROLE rolename CASCADE` instead ? We are anyway dropping the objects, why do it in two steps.
